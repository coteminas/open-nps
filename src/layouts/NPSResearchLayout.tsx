import React from 'react';
import CssBaseline from '@material-ui/core/CssBaseline';
import get from 'lodash.get';

import { GetServerSidePropsContext, GetServerSidePropsResult } from 'next';
import {
  ThemeProvider,
  createMuiTheme,
  ThemeOptions,
  makeStyles,
  WithStyles,
  createStyles,
} from '@material-ui/core/styles';

import { ITarget } from '~/model/Target';
import { IResearch } from '~/model/Research';
import { IConfig } from '~/model/Config';
import { IReviewer } from '~/model/Reviewer';
import { AddThemeOptsDefaults } from '~/util/themeOpts';

const styles = (theme) =>
  createStyles({
    root: {
      height: '100vh',
      maxWidth: '750px',
      margin: '0 auto',
      width: '100%',
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
      padding: theme.spacing(2),

      '& > h4': {
        marginBottom: theme.spacing(3),
        textAlign: 'justify',
      },
    },
  });

const useStyles = makeStyles(styles);

export interface LayoutProps extends LayoutSharedProps {
  layoutClasses: WithStyles<ReturnType<typeof styles>>['classes'];
}

interface LayoutSharedProps {
  themeOpts: ThemeOptionsConfigValues;
  templates: TemplatesConfigValues;
  researchId: string;
  data: {
    reviewer: IReviewer['meta'];
    target: IReviewer['meta'];
  };
}

interface Props extends LayoutSharedProps {
  mui: ThemeOptions;
}

interface GetServerSideProps {
  ctxResearchIdGetter(ctx: GetServerSidePropsContext): string;
  researchExtraData: { concluded: boolean };
}
export const getServerSidePropsFn = ({
  ctxResearchIdGetter,
  researchExtraData,
}: GetServerSideProps) => async (
  ctx: GetServerSidePropsContext
): Promise<GetServerSidePropsResult<Props>> => {
  const researchId = ctxResearchIdGetter(ctx);
  const { connectMongo } = await import('../util/mongo');
  const { Target, Research } = await import('../model');

  await connectMongo();

  const research: IResearch = await Research.findOne({
    _id: researchId,
    ...researchExtraData,
  }).populate('reviewer', { _id: 0 });

  const target: ITarget = research
    ? await Target.findById(research.target).populate('configs', { _id: 0 })
    : null;

  if (!research || !target) {
    return { notFound: true };
  }

  const configs: AnyObject = (target.configs as IConfig[]).reduce(
    (a, c) => ({ ...a, [c.key]: c.values }),
    {}
  );

  return {
    props: {
      mui: configs.mui,
      templates: configs.templates,
      themeOpts: AddThemeOptsDefaults(configs.themeOpts),
      data: {
        reviewer: get(research, 'reviewer.meta', {}),
        target: get(target, 'meta', {}),
      },
      researchId,
    },
  };
};

/* istanbul ignore next */
export const removeLocalJSS = (): void => {
  const jssStyles = document.querySelector('#jss-server-side');
  if (jssStyles) {
    jssStyles.parentElement.removeChild(jssStyles);
  }
};

export const withLayout = (
  Component: React.FC<LayoutProps>
): React.FC<Props> => ({ mui, ...props }): React.ReactElement => {
  const classes: LayoutProps['layoutClasses'] = useStyles();

  React.useEffect(removeLocalJSS, []);

  return (
    <ThemeProvider theme={createMuiTheme(mui)}>
      <CssBaseline />
      <Component {...props} layoutClasses={classes} />
    </ThemeProvider>
  );
};

export default withLayout;
