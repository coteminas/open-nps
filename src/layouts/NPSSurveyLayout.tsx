import React from 'react';
import CssBaseline from '@material-ui/core/CssBaseline';
import get from 'lodash.get';
import merge from 'lodash.merge';

import { GetServerSidePropsContext, GetServerSidePropsResult } from 'next';
import {
  ThemeProvider,
  createMuiTheme,
  ThemeOptions,
  makeStyles,
  WithStyles,
  createStyles,
  Theme,
} from '@material-ui/core/styles';

import { ITarget } from '~/model/Target';
import { ISurvey } from '~/model/Survey';
import { IConfig } from '~/model/Config';
import { AddThemeOptsDefaults } from '~/util/themeOpts';
import { AddDefaultTemplates } from '~/util/templates';

const childrenStyles = ({
  SurveyBoxBorderSize,
  SurveyBoxBorderRadius,
}: ThemeOptionsConfigValues) => (theme: Theme) =>
  createStyles({
    root: {
      maxWidth: '750px',
      margin: '0 auto',
      width: '100%',
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
      padding: theme.spacing(4),
      border: `${SurveyBoxBorderSize} solid ${theme.palette.primary.main}`,
      borderRadius: SurveyBoxBorderRadius,
    },
    corePhrase: {
      marginBottom: theme.spacing(3),
      textAlign: 'justify',
    },
    commentPhrase: {
      marginTop: theme.spacing(2),
      marginBottom: theme.spacing(1),
      textAlign: 'justify',
    },
    brand: {
      width: '100%',
      marginBottom: theme.spacing(3),

      '& > img': {
        width: '100%',
      },
    },
  });

const styles = (theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
      height: '100vh',
      padding: theme.spacing(2),
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
    },
  });

const useStyles = makeStyles(styles);

export interface LayoutProps extends LayoutSharedProps {
  layoutClasses: WithStyles<ReturnType<typeof childrenStyles>>['classes'];
}

interface LayoutSharedProps {
  themeOpts: ThemeOptionsConfigValues;
  templates: TemplatesConfigValues;
  surveyId: string;
  data: {
    reviewer: AnyObject;
    target: ITarget['meta'];
  };
}

interface Props extends LayoutSharedProps {
  mui: ThemeOptions;
}

interface GetServerSideProps {
  ctxSurveyIdGetter(ctx: GetServerSidePropsContext): string;
  surveyExtraData: { concluded: boolean };
}

const mapConfigsFromList = (configs: IConfig[]) =>
  configs.reduce((a, c) => ({ ...a, [c.key]: c.values }), {});

export const getServerSidePropsFn = ({
  ctxSurveyIdGetter,
  surveyExtraData,
}: GetServerSideProps) => async (
  ctx: GetServerSidePropsContext
): Promise<GetServerSidePropsResult<Props>> => {
  const surveyId = ctxSurveyIdGetter(ctx);
  const { connectMongo } = await import('../util/mongo');
  const { Target, Survey } = await import('../model');

  await connectMongo();

  const survey: ISurvey = await Survey.findOne({
    _id: surveyId,
    ...surveyExtraData,
  });

  const target: ITarget = survey
    ? await Target.findById(survey.target).populate('configs', { _id: 0 })
    : null;

  if (!survey || !target) {
    return { notFound: true };
  }

  const overrideConfigsList = await survey.getOverrideConfigs();
  const overrideConfigs: AnyObject[] = overrideConfigsList.map(
    mapConfigsFromList
  );

  const configs: AnyObject = [
    mapConfigsFromList(target.configs as IConfig[]),
    ...overrideConfigs,
  ].reduce((a, b) => merge(a, b), {});

  return {
    props: {
      mui: get(configs, 'mui', {}),
      templates: AddDefaultTemplates(configs.templates),
      themeOpts: AddThemeOptsDefaults(configs.themeOpts),
      data: {
        reviewer: JSON.parse(JSON.stringify(survey.reviewer)),
        target: get(target, 'meta', {}),
      },
      surveyId,
    },
  };
};

/* istanbul ignore next */
export const removeLocalJSS = (): void => {
  const jssStyles = document.querySelector('#jss-server-side');
  if (jssStyles) {
    jssStyles.parentElement.removeChild(jssStyles);
  }
};

export const withLayout = (
  Component: React.FC<LayoutProps>
): React.FC<Props> => ({ mui, ...props }): React.ReactElement => {
  const layoutClasses: LayoutProps['layoutClasses'] = makeStyles(
    childrenStyles(props.themeOpts)
  )();
  const classes = useStyles();

  React.useEffect(removeLocalJSS, []);

  return (
    <ThemeProvider theme={createMuiTheme(mui)}>
      <CssBaseline />
      <div className={classes.root}>
        <Component {...props} layoutClasses={layoutClasses} />
      </div>
    </ThemeProvider>
  );
};

export default withLayout;
