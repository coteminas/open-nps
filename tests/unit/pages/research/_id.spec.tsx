import React from 'react';

import {
  createSubmit,
  setValueForFieldInState,
  ResearchPage,
  ctxResearchIdGetter,
} from '~/pages/research/[id]';
import { NextRouter } from 'next/router';
import { shallow } from 'enzyme';
import { LayoutProps } from '~/layouts/NPSResearchLayout';
import { GetServerSidePropsContext } from 'next';

describe('/src/pages/research/[id]', () => {
  const fetch = global.fetch;

  beforeEach(() => {
    global.fetch = jest.fn();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  beforeAll(() => {
    global.window = ({
      location: {
        origin: 'foobar',
      },
    } as unknown) as Window & typeof globalThis;
  });

  afterAll(() => {
    global.fetch = fetch;
  });

  describe('createSubmit', () => {
    const router = ({ push: jest.fn() } as unknown) as NextRouter;
    const fakeData = { researchId: 'foo', note: '5', comment: 'bar' };
    const fakeEvent = ({
      preventDefault: jest.fn(),
    } as unknown) as React.FormEvent<HTMLFormElement>;
    const createResponse = (obj: AnyObject) => {
      const response = { json: jest.fn().mockResolvedValue(obj) };
      (global.fetch as jest.Mock).mockResolvedValue(response);
      return response;
    };

    const baseAsserts = (response) => {
      expect(response.json).toHaveBeenCalledTimes(1);
      expect(global.fetch).toHaveBeenCalledTimes(1);
      expect(global.fetch).toHaveBeenCalledWith(
        `${global.window.location.origin}/api/research/conclude`,
        {
          method: 'PUT',
          headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(fakeData),
        }
      );
    };

    it('should createSubmit right and trigger new route', async () => {
      const onSubmit = createSubmit(fakeData, router);
      const response = createResponse({ ok: 1 });

      await onSubmit(fakeEvent);

      baseAsserts(response);
      expect(router.push).toHaveBeenCalledTimes(1);
      expect(router.push).toHaveBeenCalledWith(
        `/research/thanks?researchId=${fakeData.researchId}`
      );
    });

    it('should createSubmit right and not trigger new route', async () => {
      const onSubmit = createSubmit(fakeData, router);
      const response = createResponse({ ok: 0 });

      await onSubmit(fakeEvent);

      baseAsserts(response);
      expect(router.push).not.toHaveBeenCalledTimes(1);
      expect(router.push).not.toHaveBeenCalledWith(
        `/research/thanks?researchId=${fakeData.researchId}`
      );
    });
  });

  describe('setValueForFieldInState', () => {
    it('should set correctly value', () => {
      const field = 'foo';
      const value = 'bar';
      const oldValue = 'fizz';
      const setState = jest.fn();
      const state = { preserve: 'fuzz', [field]: oldValue };

      setValueForFieldInState(state, setState)(field)(value);

      expect(setState).toHaveBeenCalledTimes(1);
      expect(setState).toHaveBeenCalledWith({ ...state, [field]: value });
    });
  });

  describe('Page Component (ResearchPage)', () => {
    it('should render properly', () => {
      const props = ({
        templates: {
          CoreQuestionPhrase: 'Foo Bar',
          ResearchCommentLabel: 'Fizz',
          ResearchCommentPlaceholder: 'Fuzz',
        },
        researchId: 'fake',
        themeOpts: {
          a: 1,
        },
        data: {
          b: 1,
        },
        layoutClasses: {
          root: 'root',
        },
      } as unknown) as LayoutProps;

      const wrap = shallow(<ResearchPage {...props} />);
      expect(wrap).toMatchSnapshot();
    });

    it('ctxResearchIdGetter', () => {
      const ctx = ({
        params: { id: 'foo' },
      } as unknown) as GetServerSidePropsContext;
      expect(ctxResearchIdGetter(ctx)).toBe(ctx.params.id);
    });
  });
});
