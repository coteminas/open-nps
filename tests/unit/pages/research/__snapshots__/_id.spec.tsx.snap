// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/src/pages/research/[id] Page Component (ResearchPage) should render properly 1`] = `
<form
  className="root"
  onSubmit={[Function]}
>
  <WithStyles(ForwardRef(Typography))
    component="h2"
    variant="h2"
  >
    Foo Bar
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ResearchNotes)
    selected={null}
    setValue={[Function]}
    themeOpts={
      Object {
        "a": 1,
      }
    }
  />
  <ResearchComment
    label="Fizz"
    placeholder="Fuzz"
    setValue={[Function]}
    value=""
  />
  <ResearchSubmit
    themeOpts={
      Object {
        "a": 1,
      }
    }
  >
    Enviar
  </ResearchSubmit>
</form>
`;
